### PR TITLE
[PropertyInfo] Support the list pseudo-type

### LINKED ISSUE #40142: [PropertyInfo] Extend the allowed array syntax to list<> in PHPDoc
**Description**
In the serialization process with the `symfony/serializer` it's common to define the `PhpDocExtractor` as part of the  `PropertyNormalizer`. With the `PhpDocExtractor` it's possible to have the following PHPDoc block on our properties:

```php
<?php

final class Company
{
    /**
     * @var Employee[]
     */
    public array $employees;

    /**
     * @param Employee[] $employees
     */
    public function __construct(array $employees)
    {
        $this->employees = $employees;
    }
}

final class Employee
{
    public string $firstName;
    public string $lastName;

    public function __construct(
        string $firstName,
        string $lastName
    ) {
        $this->firstName = $firstName;
        $this->lastName = $lastName;
    }
}
```

When supplying the following JSON, the objects are deserialized:

```json
{
  "employees": [
    { 
      "firstName": "Max",
      "lastName": "Mustermann"
    },
    {
      "firstName": "Clara",
      "lastName": "Carlston"
    }
  ]
}
```

An alternative Annotation would be the usage of `list<Employee>`. Which has the following advantages:

- It contains more information as `array<Employee>` (the key is an int)
- It contains more information then `array<int, Employee>` (the key has to start with 0 and has no missing steps)

`list<...>` is supported by Psalm, PHPStan and PHPStorm, but isn't by the `PhpDocExtractor`. When used with the serializer, it result's in a `ReflectionException` with `Class App\...\list does not exist`.

I searched through the logic used and found the following snippet in [PhpDocTypeHelper](https://github.com/symfony/symfony/blob/f8b154357d498f328f06f1c2d9b48a9e97a7f402/src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php#L119):

```php
if (0 === strpos($docType, 'array<') && $type instanceof Array_) {
    // array<value> is converted to x[] which is handled above
    // so it's only necessary to handle array<key, value> here
    $collectionKeyType = $this->getTypes($type->getKeyType())[0];

    $collectionValueTypes = $this->getTypes($type->getValueType());
    if (\count($collectionValueTypes) != 1) {
        // the Type class does not support union types yet, so assume that no type was defined
        $collectionValueType = null;
    } else {
        $collectionValueType = $collectionValueTypes[0];
    }

    return new Type(Type::BUILTIN_TYPE_ARRAY, $nullable, null, true, $collectionKeyType, $collectionValueType);
}
```

It's my understanding, that it would be enough to extend that with the `list` type like the following:

```php
if ((0 === strpos($docType, 'array<') || 0 === strpos($docType, 'list<')) && $type instanceof Array_) {
```

But I'm not 100% sure.

Would it be possible to extend it like this? 

If so, would be happy to supply the Pull Request for it.

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #40142
| License       | MIT
| Doc PR        | N/A

This PR teaches `PhpDocExtractor` and `PhpStanextractor` to understand the generic `list<>` type. It also does some cleanup in both extractors.

### CODE REVIEW

File: src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php [hunk 1]
```diff
@@ -32,9 +32,9 @@
  */
 final class PhpStanExtractor implements PropertyTypeExtractorInterface, ConstructorArgumentTypeExtractorInterface
 {
-    public const PROPERTY = 0;
-    public const ACCESSOR = 1;
-    public const MUTATOR = 2;
+    private const PROPERTY = 0;
+    private const ACCESSOR = 1;
+    private const MUTATOR = 2;
```
Comment (derrabus) [inline src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php:37]: I don't think that those constants should be public. They're only used internally.

File: src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php [hunk 2]
```diff
@@ -91,7 +92,13 @@ private function createType(DocType $type, bool $nullable, string $docType = nul
         $docType = $docType ?? (string) $type;
 
         if ($type instanceof Collection) {
-            [$phpType, $class] = $this->getPhpTypeAndClass((string) $type->getFqsen());
+            $fqsen = $type->getFqsen();
+            if (!class_exists(List_::class) && $fqsen && 'list' === $fqsen->getName() && !class_exists((string) $fqsen)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php]: could the class_exists check be in a tight loop? if yes, we should move them somewhere else or at least as last conditions

Comment (derrabus) [reply src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php]: I'll change the order of the conditions.

Comment (derrabus) [reply src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php]: Done. I've also made sure we're not triggering the autoloader inside the loop.

File: src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php [hunk 3]
```diff
@@ -19,6 +20,10 @@
 use phpDocumentor\Reflection\Types\Nullable;
 use Symfony\Component\PropertyInfo\Type;
 
+// Workaround for phpdocumentor/type-resolver < 1.6
+// We trigger the autoloader here, son we don't need to trigger it inside the loop later.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php]: s/son/so

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php (hunk 3 above)
Comment (derrabus) [reply src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php]: ✅

**Pull request merged**